lib/api/core/validation/index.js

```diff
@@ -0,0 +1,788 @@
+    }
+  }).then(documentBody => {
+    var
+      validation = true,
+      strictnessError = false,
+      errorMessages = structuredError ? {} : [],
+      /** @type {CollectionSpecification} */
+      collectionSpec;
+
+    if (this.specification && this.specification[index] && this.specification[index][collection]) {
+      collectionSpec = this.specification[index][collection];
+
+      if (collectionSpec.fields) {
+        try {
+          validation = validation && this.recurseFieldValidation(documentBody, updateBodyClone || null, collectionSpec.fields.children, collectionSpec.strict, errorMessages, structuredError);
+        }
+        catch (error) {
+          if (error.message === 'strictness') {
+            // The strictness message can be received here only if it happens at the validation of the document's root
+            addErrorMessage('document', errorMessages, 'The document validation is strict; it can not add unspecified sub-fields.', structuredError);
+            strictnessError = true;
+          }
+        }
+
+        if (!strictnessError && !validation) {
+          addErrorMessage('document', errorMessages, 'The document is not valid due to field validation.', structuredError);
+        }
+      }
+
+      if (collectionSpec.validators) {
+        return this.dsl.test(index, collection, documentBody, id)
+          .then(filters => {
+            var result = false;
+
+            filters.forEach(filter => {
+              if (filter === collectionSpec.validators) {
+                result = true;
+                return false;
+              }
+            });
+
+            if (!result) {
+              validation = false;
+              addErrorMessage('document', errorMessages, 'The document does not match validation filters.', structuredError);
+            }
+
+            return {
+              documentBody: updateBodyClone || requestBodyClone,
+              errorMessages: errorMessages,
+              validation: validation
+            };
+          });
+      }
+    }
+
+    return {
+      documentBody: updateBodyClone || requestBodyClone,
+      errorMessages: errorMessages,
+      validation: validation
+    };
+  });
+};
+
+/**
+ * @param {*} documentSubset
+ * @param {*} originalBodySubset
+ * @param {StructuredFieldSpecification} collectionSpecSubset
+ * @param {boolean} strictness
+ * @param {string[]} errorMessages
+ * @param {boolean} structuredError
+ */
+Validation.prototype.recurseFieldValidation = function (documentSubset, originalBodySubset, collectionSpecSubset, strictness, errorMessages, structuredError) {
+  if (strictness) {
+    if (!checkAllowedProperties(documentSubset, Object.keys(collectionSpecSubset))) {
+      // We use a throw to be able to provide information about the field or the document in whole
+      throw new BadRequestError('strictness');
+    }
+  }
+
+  return Object.keys(collectionSpecSubset).reduce((reductionResult, fieldName) => {
```

Comment:
Just a reminder, after we debate about fail-fast vs. get all errors: if the former is chosen, then Array.every should be used instead of Array.reduce.